firedrake/adjoint_utils/function.py

```diff
@@ -225,23 +225,32 @@ def _ad_convert_riesz(self, value, options=None):
         from firedrake import Function, Cofunction
 
         options = {} if options is None else options
-        riesz_representation = options.get("riesz_representation", "l2")
+        riesz_representation = options.get("riesz_representation", "L2")
         solver_options = options.get("solver_options", {})
         V = options.get("function_space", self.function_space())
 
-        if riesz_representation != "l2" and not isinstance(value, Cofunction):
-            raise TypeError("Expected a Cofunction")
-        elif not isinstance(value, (Number, Cofunction, Function)):
+        if not isinstance(value, (Number, Cofunction, Function, Number)):
             raise TypeError("Expected a Cofunction, Function or a float")
 
-        if riesz_representation == "l2":
-            if isinstance(value, (Cofunction, Function)):
-                return Function(V, val=value.dat)
-            else:
+        if isinstance(value, Number):
+            if value == 0.:
+                # l2 Riesz map is directly applied when the value is a real number 0..
+                # This is seen in adjoint-based derivative when the functional
+                # is independent of the control variable.
+                return Function(V)
+            elif self.ufl_element().family() == "Real":
+                # Apply the l2 Riesz map for the case where self is a function in Real space.
                 f = Function(V)
-                with stop_annotating():
-                    f.assign(value)
+                f.assign(value)
                 return f
+            else:
+                raise TypeError("Riesz map of a non-zero scalar is not supported for non-Real function spaces.")
+
+        if riesz_representation == "l2":
+            return Function(V, val=value.dat)
+
+        if riesz_representation != "l2" and not isinstance(value, Cofunction):
```

Comment:
`riesz_representation != "l2"` will always be `True` at this point, no?

Reply:
Yes. You are right. I just rewrote `if not isinstance(value, Cofunction)`.